[AutoDiff] Add reverse-mode `switch_enum` support for `Optional`.

efremale opened the pull request:
Reverse mode autodiff now works on `Optional`s, i.e. `switch_enum` and `switch_enum_addr` SIL instructions are supported now.

dan-zheng committed:
[WIP] Start optional differentiation support.

dan-zheng committed:
[WIP]

efremale committed:
[AutoDiff] Add `Optional` support to reverse mode

Reverse mode autodiff now works on `Optional`s, i.e. `switch_enum` SIL instruction is
supported now. The `switch_enum_addr` instruction support is still missing.

efremale committed:
[AutoDiff] Add reverse-mode differentiation support for `switch_enum_addr`

On top of switch_enum, now switch_enum_addr SIL instruction can be differentiated.

efremale commented:
cc @dan-zheng

efremale committed:
Code cleanup

efremale committed:
Fix Differentialbe conformance lookup: use `T` instead of `T.TangentVector`

dan-zheng commented on test/AutoDiff/SILOptimizer/optional.swift:
```diff
@@ -0,0 +1,75 @@
+//===----------------------------------------------------------------------===//
+
+/*
+// Note: this lowers to `try_apply` instead of `switch_enum`, and is
```
Please reference TF-433, which tracks `try_apply` differentiation:

```suggestion
// TODO(TF-433): this lowers to `try_apply` instead of `switch_enum`, and is
```

dan-zheng commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -2094,6 +2099,85 @@ void PullbackCloner::Implementation::visitSILBasicBlock(SILBasicBlock *bb) {
+          auto *optionalEnumDecl = getASTContext().getOptionalDecl();
+          auto operandTy = sei->getOperand()->getType();
+          if (operandTy.getASTType().getEnumOrBoundGenericEnum() == optionalEnumDecl) {
+            // `Optional<T>`
+            auto optionalTy = remapType(operandTy);
+            // `T`
+            auto Ty = optionalTy.getOptionalObjectType();
+            // `T.TangentVector`
+            auto tanTy = remapType(concreteBBArgAdjCopy->getType());
+            // `Optional<T.TangentVector>`
+            auto optionalTanTy = SILType::getOptionalType(tanTy);
+            // `Optional<T>.TangentVector`
+            auto tangentVectorTy = getRemappedTangentType(optionalTy);
```
The goal here seems to be to get the `Optional<T>.TangentVector` type. Could you collapse these into a single call to `getRemappedTangentType`?

```suggestion
            // `Optional<T>.TangentVector`
            auto optionalTanTy = getRemappedTangentType(operandTy);
```

efremale replied:
The goal is to get two things:
1. `Optional<T>.TangentVector`
2. `Optional<T.TangentVector>` (for a parameter to pass to`Optional<T>.TangentVector.init`)
The `getRemappedTangentType` gives 1., but I don't see a straightforward way to shorten the rest of the code that gets 2.

dan-zheng replied:
Ah, I see. Both types seem necessary.

dan-zheng committed:
Minor cleanup.

Edit comments, run `git clang-format`.

dan-zheng committed:
Move e2e optional differentiation to validation-test subdirectory.

dan-zheng committed:
More cleanup, expose failing test.

Move `Optional` adjoint value propagation logic to a dedicated helper method.
Add failing test case involving nested `Optional` differentiation.

rxwei reviewed (commented):
Is there a reason why `Optional` is being specially treated rather than supported via differentiating `switch_enum` on any differentiable enum type?

efremale committed:
Fix adjoint propagation for adjoints of address-only value category, add nested `Optional`s tests, code cleanup

efremale committed:
Code cleanup, add more generic tests

efremale committed:
Code cleanup, add doc comment for `propagateAdjoingForOptional` function

efremale committed:
Tests cleanup

efremale committed:
Fix nested Optional tests to avoid a crash, add more nested tests

efremale committed:
Cleanup

dan-zheng commented:
> Is there a reason why `Optional` is being specially treated rather than supported via differentiating `switch_enum` on any differentiable enum type?

No strong reason, other than to start progress on `Optional` differentiation until a general `switch_enum` differentiation approach is clear.

We discussed a general `switch_enum` differentiation approach offline, which seems promising. How about landing this PR with special-case `Optional` support in the meantime? The `Optional` differentiation test cases are valuable and should continue to pass if we generalize the approach.

dan-zheng committed:
NFC: clean up.

dan-zheng committed:
Expose crasher involving `Tracked<Float>?` differentiation.

efremale committed:
Add `Tracked<Float>` testcases

Expose extra release error, add a temporary workaround.

dan-zheng committed:
NFC: clean up.

rxwei commented:
> We discussed a general `switch_enum` differentiation approach offline, which seems promising. How about landing this PR with special-case `Optional` support in the meantime? The `Optional` differentiation test cases are valuable and should continue to pass if we generalize the approach.

Sure.

dan-zheng committed:
NFC: clean up.

dan-zheng commented:
@swift-ci Please test

rxwei commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -1972,6 +2014,106 @@ void PullbackCloner::Implementation::emitZeroDerivativesForNonvariedResult(
+  addToAdjointBuffer(bb, optionalValue, optTanAdjBuf, pbLoc);
+  // TODO: Investigate why enabling this produces a crash for `Tracked<Float>?`
+  // differentiation.
+  // builder.emitDestroyAddr(pbLoc, optTanAdjBuf);
```
Can we figure out what's going on with this before merging?  Why is it a crasher rather than an ownership verification error?

dan-zheng replied:
> Why is it a crasher rather than an ownership verification error?

I have the same question.

Uncommenting this `builder.emitDestroyAddr(pbLoc, optTanAdjBuf)` causes the test to fail with an "extra release" error at runtime, which seems unexpected because `optTanAdjBuf` is a locally created `alloc_stack`.

```
[ RUN      ] OptionalDifferentiation.Let
stderr>>> Assertion failed: (getStrongExtraRefCount() >= dec && "releasing reference whose refcount is already zero"), function decrementStrongExtraRefCount, file /Users/danielzheng/swift-merge/swift/include/swift/Runtime/../../../stdlib/public/SwiftShims/RefCount.h, line 567.
stderr>>> CRASHED: SIGABRT
```

Interestingly, commenting `builder.emitDestroyAddr` doesn't appear to cause memory leaks when running the test (according to `valgrind`).

rxwei replied:
Can you dump the generated raw SIL of both cases and let me take a look?

dan-zheng replied:
Yes, thanks for offering to look.

[Here's a crash reproducer](https://gist.github.com/dan-zheng/f49f5e3def048c786c453e05f657d008), with SIL for both cases and full `-Xllvm debug-only=differentiation` output.

rxwei replied:
Here's the issue.  You called `Optional.TangentVector.init(_:)` with argument `%107` but you had a `destroy_addr %107` later in the code.  This itself is incorrect: `%107` should be considered consumed after the call.

```swift
  %107 = alloc_stack $Optional<Tracked<Float>>    // users: %115, %114, %113, %110, %108
  %108 = init_enum_data_addr %107 : $*Optional<Tracked<Float>>, #Optional.some!enumelt // user: %109
  copy_addr %3 to [initialization] %108 : $*Tracked<Float> // id: %109
  inject_enum_addr %107 : $*Optional<Tracked<Float>>, #Optional.some!enumelt // id: %110
  // function_ref Optional<A>.TangentVector.init(_:)
  %111 = function_ref @$sSq16_DifferentiationAA14DifferentiableRzlE13TangentVectorVyADyx_GACQzSgcfC : $@convention(method) <τ_0_0 where τ_0_0 : Differentiable> (@in Optional<τ_0_0.TangentVector>, @thin Optional<τ_0_0>.TangentVector.Type) -> @out Optional<τ_0_0>.TangentVector // user: %113
  %112 = metatype $@thin Optional<Tracked<Float>>.TangentVector.Type // user: %113
  %113 = apply %111<Tracked<Float>>(%106, %107, %112) : $@convention(method) <τ_0_0 where τ_0_0 : Differentiable> (@in Optional<τ_0_0.TangentVector>, @thin Optional<τ_0_0>.TangentVector.Type) -> @out Optional<τ_0_0>.TangentVector
BAD! --->  destroy_addr %107 : $*Optional<Tracked<Float>>  // id: %114
  dealloc_stack %107 : $*Optional<Tracked<Float>> // id: %115
```

The `destroy_value %107` did not happen to crash because coincidentally `TangentVector.init(_:)` is simply wrapping `%107` in a `TangentVector` wrapper. When you `destroy_value %107`, it released the underlying value. Then when you `destroy_value %106`, which at this point still wraps `%107`'s value, it's a double-free.

rxwei replied:
@gottesmm Any idea why the ownership verifier didn't catch this?
```console
%107 = alloc_stack $Optional<Tracked<Float>>
%113 = apply %111<Tracked<Float>>(%106, %107, %112) : $@convention(method) <τ_0_0 where τ_0_0 : Differentiable> (@in Optional<τ_0_0.TangentVector>, @thin Optional<τ_0_0>.TangentVector.Type) -> @out Optional<τ_0_0>.TangentVector
**Use after consume**: destroy_addr %107 : $*Optional<Tracked<Float>>  // id: %114
```

dan-zheng replied:
Thanks. That makes sense, because `Optional<A>.TangentVector.init(_:)` consumes its `@in` parameter. Fixed.

dan-zheng committed:
NFC: clean up.

dan-zheng commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -2110,18 +2253,64 @@ void PullbackCloner::Implementation::visitSILBasicBlock(SILBasicBlock *bb) {
+    // an `Optional`-typed value. `switch_enum` instructions require
+    // special-case adjoint value propagation for the operand.
+    auto isSwitchEnumInstOnOptional =
+        [&Ctx = getASTContext()](TermInst *termInst) {
```
Done.

dan-zheng committed:
Fix `destroy_addr` double-free error.

`Optional.TangentVector.init` takes a `@in` argument address.

Do not `destroy_addr` the argument address after applying
`Optional.TangentVector.init`.

dan-zheng commented:
@swift-ci Please test
